projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/control/LLVMLoopDispatchNode.java

```diff
@@ -0,0 +1,265 @@
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMControlFlowNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMStatementNode;
+
+public final class LLVMLoopDispatchNode extends LLVMExpressionNode {
+
+    private final FrameSlot exceptionValueSlot;
+    private final int headerId;
+    @Children private final LLVMStatementNode[] bodyNodes;
+    @CompilationFinal private final Integer[] indexMapping;
+    @CompilationFinal(dimensions = 2) private final FrameSlot[][] beforeBlockNuller;
+    @CompilationFinal(dimensions = 2) private final FrameSlot[][] afterBlockNuller;
+    @CompilationFinal private final Integer[] loopSuccessors;
+    @CompilationFinal private final FrameSlot successorSlot;
+
+    public LLVMLoopDispatchNode(FrameSlot exceptionValueSlot, LLVMStatementNode[] bodyNodes, FrameSlot[][] beforeBlockNuller, FrameSlot[][] afterBlockNuller, int headerId, Integer[] indexMapping, Integer[] successors, FrameSlot successorSlot) {
+        this.exceptionValueSlot = exceptionValueSlot;
+        this.bodyNodes = bodyNodes;
+        this.beforeBlockNuller = beforeBlockNuller;
+        this.afterBlockNuller = afterBlockNuller;
+        this.indexMapping = indexMapping;
+        this.headerId = headerId;
+        this.loopSuccessors = successors;
+        this.successorSlot = successorSlot;
+    }
+
+    @ExplodeLoop
+    private boolean isInLoop(int bci) {
+        for(int i : loopSuccessors) {
+            if(i == bci)
+                return false;
+        }
+        return true;
+    }
+
+    @Override
+    @ExplodeLoop(kind = LoopExplosionKind.MERGE_EXPLODE)
+    public Object executeGeneric(VirtualFrame frame) {
+        boolean ret = false; // helper variable to not return at first encounter of headerId
+
+        CompilerAsserts.compilationConstant(bodyNodes.length);
+        int basicBlockIndex = headerId;
+
+        outer: while (isInLoop(basicBlockIndex) && (basicBlockIndex != headerId || !ret)) { // do-while loop fails at PE
+            CompilerAsserts.partialEvaluationConstant(basicBlockIndex);
+
+            ret = true;
+
+            if(bodyNodes[indexMapping[basicBlockIndex]] instanceof LLVMLoopNode) {
+                LLVMLoopNode loop = (LLVMLoopNode) bodyNodes[indexMapping[basicBlockIndex]];
+                loop.execute(frame);
+
+                Integer[] successors = loop.getSuccessors();
+                for(int i = 0; i < successors.length; i++) {
+                    try {
+                        if(frame.getInt(successorSlot) == successors[i]) {
+                            basicBlockIndex = successors[i];
+                            continue outer;
+                        }
+                    } catch (FrameSlotTypeException e) {
+                        CompilerDirectives.transferToInterpreter();
+                        throw new RuntimeException("Error while reading from loop successor frame slot - type mismatch!");
```

Comment:
I already mentioned this above, but a more descriptive error message would come in handy for debugging. In general, since this class is in large parts a duplicate of `LLVMDispatchBasicBlockNode` it would be preferable to have the common parts in a common baseclass, but it may be hard to accomplish this without any performance regressions. Perhaps you could look into this as future work.